code/datums/mapgen/CaveGenerator.dm

```diff
@@ -36,107 +50,114 @@
 
 /datum/map_generator/cave_generator/New()
 	. = ..()
-	if(!mob_spawn_list)
-		mob_spawn_list = list(/mob/living/simple_animal/hostile/asteroid/goldgrub = 1, /mob/living/simple_animal/hostile/asteroid/goliath = 5, /mob/living/simple_animal/hostile/asteroid/basilisk = 4, /mob/living/simple_animal/hostile/asteroid/hivelord = 3)
-	if(!megafauna_spawn_list)
-		megafauna_spawn_list = GLOB.megafauna_spawn_list
-	if(!flora_spawn_list)
-		flora_spawn_list = list(/obj/structure/flora/ash/leaf_shroom = 2 , /obj/structure/flora/ash/cap_shroom = 2 , /obj/structure/flora/ash/stem_shroom = 2 , /obj/structure/flora/ash/cacti = 1, /obj/structure/flora/ash/tall_shroom = 2, /obj/structure/flora/ash/seraka = 2)
-	if(!feature_spawn_list)
-		feature_spawn_list = list(/obj/structure/geyser/random = 1)
-
-/datum/map_generator/cave_generator/generate_terrain(list/turfs)
+	if(!weighted_mob_spawn_list)
+		weighted_mob_spawn_list = list(/mob/living/simple_animal/hostile/asteroid/goldgrub = 1, /mob/living/simple_animal/hostile/asteroid/goliath = 5, /mob/living/simple_animal/hostile/asteroid/basilisk = 4, /mob/living/simple_animal/hostile/asteroid/hivelord = 3)
+	mob_spawn_list = expand_weights(weighted_mob_spawn_list)
+	mob_spawn_no_mega_list = expand_weights(weighted_mob_spawn_list - SPAWN_MEGAFAUNA)
+	if(!weighted_megafauna_spawn_list)
+		weighted_megafauna_spawn_list = GLOB.megafauna_spawn_list
+	megafauna_spawn_list = expand_weights(weighted_megafauna_spawn_list)
+	if(!weighted_flora_spawn_list)
+		weighted_flora_spawn_list = list(/obj/structure/flora/ash/leaf_shroom = 2 , /obj/structure/flora/ash/cap_shroom = 2 , /obj/structure/flora/ash/stem_shroom = 2 , /obj/structure/flora/ash/cacti = 1, /obj/structure/flora/ash/tall_shroom = 2, /obj/structure/flora/ash/seraka = 2)
+	flora_spawn_list = expand_weights(weighted_flora_spawn_list)
+	if(!weighted_feature_spawn_list)
+		weighted_feature_spawn_list = list(/obj/structure/geyser/random = 1)
+	feature_spawn_list = expand_weights(weighted_feature_spawn_list)
+	open_turf_types = expand_weights(weighted_open_turf_types)
+	closed_turf_types = expand_weights(weighted_closed_turf_types)
+
+
+/datum/map_generator/cave_generator/generate_terrain(list/turfs, area/generate_in)
 	. = ..()
+	if(!(generate_in.area_flags & CAVES_ALLOWED))
+		return
+
 	var/start_time = REALTIMEOFDAY
 	string_gen = rustg_cnoise_generate("[initial_closed_chance]", "[smoothing_iterations]", "[birth_limit]", "[death_limit]", "[world.maxx]", "[world.maxy]") //Generate the raw CA data
 
+	// Area var pullouts to make accessing in the loop faster
+	var/flora_allowed = generate_in.area_flags & FLORA_ALLOWED && length(flora_spawn_list)
+	var/feature_allowed = generate_in.area_flags & FLORA_ALLOWED && length(feature_spawn_list)
+	var/mobs_allowed = generate_in.area_flags & MOB_SPAWN_ALLOWED && length(mob_spawn_list)
+	var/megas_allowed = generate_in.area_flags & MEGAFAUNA_SPAWN_ALLOWED && length(megafauna_spawn_list)
```

Comment:
```suggestion
	var/flora_allowed = (generate_in.area_flags & FLORA_ALLOWED) && length(flora_spawn_list)
	var/feature_allowed = (generate_in.area_flags & FLORA_ALLOWED) && length(feature_spawn_list)
	var/mobs_allowed = (generate_in.area_flags & MOB_SPAWN_ALLOWED) && length(mob_spawn_list)
	var/megas_allowed = (generate_in.area_flags & MEGAFAUNA_SPAWN_ALLOWED) && length(megafauna_spawn_list)
```

Reply:
why?

Reply:
it looks like it'll easily confuse people between uses of & and &&, or at the very least cause bad coding practices which could cause errors for people who reference off of this

Reply:
the weak should fear the strong, but ok

Reply:
I also don't really see how this is bad practice

Reply:
we do tend to wrap these in parens

Reply:
aight, if you say so